lib/posix/pthread_key.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <misc/dlist.h>
+#include <semaphore.h>
+#include <errno.h>
+#include <posix/pthread.h>
+
+static sem_t sem;
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *newKey;
+
+	*key = NULL;
+
+	sem_init(&sem, 0, 1);
+
+	newKey = (pthread_key_obj *)k_malloc(sizeof(pthread_key_obj));
+
+	if (newKey == NULL) {
+		errno = ENOMEM;
+		return -1;
+	}
+
+	sys_dlist_init(&(newKey->KeyDataL));
+
+	newKey->destructor = destructor;
+	*key = (void *)newKey;
+
+	return 0;
+}
+
+/**
+ * @brief Delete a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
+{
+	pthread_key_obj *keyObj = (pthread_key_obj *)key;
+	KeyData *keyData;
+
+	sem_wait(&sem);
+
+	/* Delete thread-specific elements associated with the key */
+	while (!sys_dlist_is_empty(&(keyObj->KeyDataL))) {
+
+		/* Remove the object from the list KeyDataL */
+		keyData = (KeyData *)sys_dlist_get(&(keyObj->KeyDataL));
```

Comment:
Casting from a `void *` to any other pointer in C without a cast is valid.  Please remove all casts that are unnecessary.

Reply:
Okay, if I remove them, they work, but here, `sys_dlist_get()`'s return type is `sys_dnode_t *`, and not `void *`, so when compiling, it complains of a `incompatible pointer types`. Do I still remove it?

Reply:
I will remove the cast in other places, like in `k_malloc` calls